gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+                                    "invalid_sensor"
+                                }
+                                MeasurementError::NoReading => "no_reading",
+                                MeasurementError::NotPresent => "not_present",
+                                MeasurementError::DeviceError => "device_error",
+                                MeasurementError::DeviceUnavailable => {
+                                    "device_unavailable"
+                                }
+                                MeasurementError::DeviceTimeout => {
+                                    "device_timeout"
+                                }
+                                MeasurementError::DeviceOff => "device_off",
+                            };
+                            let datum = sensor_errors
+                                .entry(SensorErrorKey {
+                                    name: sensor.clone(),
+                                    kind,
+                                    error,
+                                })
+                                .or_insert(Cumulative::new(0));
+                            // TODO(eliza): perhaps we should treat this as
+                            // "level-triggered" and only increment the counter
+                            // when the sensor has *changed* to an errored
+                            // state after we have seen at least one good
+                            // measurement from it since the last time the error
+                            // was observed?
+                            datum.increment();
+                            Sample::new(
+                                target,
+                                &metric::SensorErrorCount {
+                                    error: Cow::Borrowed(error),
+                                    sensor,
+                                    datum: *datum,
+                                    sensor_kind: Cow::Borrowed(kind),
+                                },
+                            )
+                        }
+                    };
+                    try_sample!(sample);
+                }
+            }
+
+            // Now, fetch the SP's state *again*. It is possible that, while we
+            // were scraping the SP's samples, the SP's identity changed in some
+            // way: perhaps its version was updated during the poll, or it
+            // was removed from the rack and replaced with an entirely different
+            // chassis! If that's the case, some of the samples we collected may
+            // have a metrics target describing the wrong thing (e.g. they could
+            // still have the previous firmware's `hubris_archive_id`, if the SP
+            // was updated). In that case, we need to throw away the samples we
+            // collected and try again, potentially rebuilding our understanding
+            // of the SP's inventory.
```

Comment:
Would there ever be a case where this data would still be meaningful?

Reply:
The measurements themselves are still meaningful, but they will be labeled with an incorrect metrics target --- they'll either have a wrong Hubris archive version, if a firmware update occurred mid-poll, which is not great; or a wrong chassis serial/model number/revision, if a sled was replaced with a different sled mid-scrape, which is *worse*. We will have published measurements from the previous target one second ago, and we'll immediately poll again if we see that the SP's state has changed, so I don't think throwing away the incorrectly-labled samples is all that critical of a loss of data. 

@jgallagher and I did discuss some alternative solutions, such as having the poller task ask the SP for its state after *every* `ComponentDetails` request, or including the Hubris firmware ID in the `ComponentDetails` message from the SP (see https://github.com/oxidecomputer/omicron/pull/6354#issuecomment-2302942594). The first approach means a lot of additional requests to the SP, which seems unfortunate, as it would increase management network traffic and SP load. The other approach requires changes to the gateway-sp-comms protocol. But, these strategies would allow us to keep the samples collected with a previous firmware revision when an update occurs mid-scrape, because we know *precisely* when it changed, rather than performing the whole scrape and then checking. 

I think the solution of reading the whole SP state on every poll is probably a non-starter. It might be worth pursuing the approach of including the firmware version on all messages, in the future --- I might work on that once this change merges. For now, I think the approach of optimistically performing the whole scrape and then checking if the firmware version is still the same is the best compromise between not recording wrong data and not weighing down the wire protocol with a bunch of metadata for every component we read data from. What do you think?

Reply:
Instead of including firmware version specifically, what about letting MGS send a (randomly generated on the host) `u64` to the SP, then include that same value in every message (and return an error if it doesn't match) – basically the dumbest possible shared secret?

This would definitely be an MGS change (and an annoying one at that), so I'm fine with the current implementation of asking for the SP state again.

Reply:
Hmm, yeah, adding a way to set a nonce to the gateway-sp-comms protocol could be a good idea. It would also detect cases where the `control-plane-agent` task has been *restarted*, which...is either good or bad. I'd like to look into this as follow-up work.